sw/device/lib/drivers/gpio/gpio.c

```diff
@@ -1,213 +1,525 @@
+******************************* C SOURCE FILE *****************************
+**                            *******************                                                                                               
+**                                                                         
+** project  : X-HEEP                                                       
+** filename : gpio.c                                                                                                        
+**                                                                         
+***************************************************************************
+**                                                                         
+** Copyright (c) EPFL contributors.                                        
+** All rights reserved.                                                    
+**                                                                         
+***************************************************************************
+*/
+
+/***************************************************************************/
+/***************************************************************************/
+/**
+* @file     gpio.c
+* @date     30/03/2023
+* @author   Hossein taji
+* @version  1
+* @brief    GPIO driver 
+*/
+
+/****************************************************************************/
+/**                                                                        **/
+/*                             MODULES USED                                 */
+/**                                                                        **/
+/****************************************************************************/
 
 #include "gpio.h"
-
 #include "gpio_regs.h"  // Generated.
+#include "gpio_structs.h"
+#include "bitfield.h"
+#include "x-heep.h"
+
+/****************************************************************************/
+/**                                                                        **/
+/*                        DEFINITIONS AND MACROS                            */
+/**                                                                        **/
+/****************************************************************************/
 
 /**
- * Gives the mask that corresponds to the given bit index.
- *
- * @param index Bit index in a 32-bit register.
+ * Mask for different bit field width
  */
-static uint32_t index_to_mask(uint32_t index) { return 1u << index; }
+#define MASK_1b     0b1
+#define MASK_2b     0b11
 
-gpio_result_t gpio_init(gpio_params_t params, gpio_t *gpio) {
-  if (gpio == NULL) {
-    return kGpioBadArg;
-  }
+/**
+ * GPIO_EN values 
+ */
+#define GPIO_EN__ENABLED     1
+#define GPIO_EN__DISABLED    0
+
+/**
+ * GPIO_CLEAR and GPIO_SET putting and removing mask
+ */
+#define GPIO_CLEAR__PUT_MASK 1
+#define GPIO_CLEAR__REMOVE_MASK 0
+#define GPIO_SET__PUT_MASK 1
+#define GPIO_SET__REMOVE_MASK 0
```

Comment:
do not duplicate defines, just:
`#define GPIO_PUT_MASK 0x01
#define GPIO_REMOVE_MASK 0x00`